### PR TITLE
Search results history feature

### PR DESCRIPTION
**Pull request by bogdasar1985**
Implementation of https://github.com/newsboat/newsboat/issues/1785

### CODE REVIEW

**Comment by Minoru** [comment]
Works quite nicely, thanks a lot for the work!

I like this solution in principle, but there's a decision I disagree with: there should be no `OP_NEXTSEARCHRESULT`. It reintroduces the "tree of searches" problem I described in https://github.com/newsboat/newsboat/issues/1785#issuecomment-1089268515. I think it brings more confusion than utility, so I'd remove this operation altogether.

I have more comments, but they're lower-level, so let's agree on this point first.

**Comment by Minoru** [comment]
Also, please don't merge other branches into PRs — it makes the history "gnarly", harder to read and slower to bisect. If I *really* needed this to be updated, I'd ask you to rebase instead.

**Comment by bogdasar1985** [comment]
@Minoru ok, I will delete `OP_NEXTSEARCHRESULTS`. And I think we should add a hint about `q` is Quit key.

**Comment by Minoru** [comment]
Cool! But now `prevsearchresults` is kind of broken:

1. search for something;
2. refine the search;
3. press <kbd>z</kbd> to go back to the first results set;
4. refine the search again;
5. press <kbd>z</kbd> to go back to the first results set.

What actually happens on the last step is that we go back to the results shown on step 2.

I think `searchistorypos` can be removed now, and `searchresultshistory` can be treated as a stack: new searches push onto it, `prevsearchresults` pop off it.

**Comment by bogdasar1985** [comment]
@Minoru Strange, I can't reproduce the bug. It seems it depends from feed and search requests, but shouldn't. But I really agree that `std::stack` will be better here in any case.

**Comment by Minoru** [comment]
Hmm, yeah, I can't reproduce it either =\ Well, anyway, the stack fixes it, so all is well. I will now make a proper code review of this PR, stay tuned :)

File: include/searchresultsformaction.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#include "logger.h"
+#include "regexmanager.h"
+#include "rssfeed.h"
+#include "view.h"
+#include "keymap.h"
+#include <stack>
+
+namespace newsboat {
+
+class SearchResultsFormAction : public ItemListFormAction {
```
Comment (Minoru) [inline include/searchresultsformaction.h:14]: It's a bit confusing that the class is named "SearchResults…" when other constants say "search results **list**". Can this be renamed to `SearchResultsListFormAction`? (Don't forget to rename the files and the include guards too!)

It's a long name, but at least it's consistent with all the other stuff we already have.

File: include/searchresultsformaction.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
+#ifndef NEWSBOAT_SEARCHRESULTSFORMACTION_H_
+#define NEWSBOAT_SEARCHRESULTSFORMACTION_H_
+#include "configcontainer.h"
+#include "itemlistformaction.h"
+#include "logger.h"
```
Comment (Minoru) [inline include/searchresultsformaction.h:5]: Do we need logger.h in here? It doesn't look like this header logs anything.

File: src/itemlistformaction.cpp [hunk 3]
```diff
@@ -915,10 +915,6 @@ void ItemListFormAction::qna_start_search()
 	std::shared_ptr<RssFeed> search_dummy_feed(new RssFeed(rsscache, ""));
 	search_dummy_feed->set_search_feed(true);
 	search_dummy_feed->add_items(items);
-
-	if (show_searchresult) {
```
Comment (Minoru) [inline src/itemlistformaction.cpp:919]: Is `show_searchresult` even needed now that we have a separate class for search results? I wonder what other stuff can be removed from `ItemListFormAction` now, or at least moved down into the new class? (If it turns out there's a lot of stuff, it better be postponed to a separate PR).

File: src/keymap.cpp [hunk 4]
```diff
@@ -80,87 +80,87 @@ static const std::vector<OpDesc> opdescs = {
...
-	{OP_SAVEALL, "save-all", "", translatable("Save articles"), KM_ARTICLELIST},
+	{OP_SAVE, "save", "s", translatable("Save article"), KM_ARTICLELIST | KM_ARTICLE | KM_SEARCHRESULTSLIST},
+	{OP_SAVEALL, "save-all", "", translatable("Save articles"), KM_ARTICLELIST | KM_SEARCHRESULTSLIST},
 	{
 		OP_NEXT,
 		"next",
 		"J",
 		translatable("Go to next entry"),
-		KM_FEEDLIST | KM_ARTICLELIST | KM_ARTICLE | KM_DIALOGS | KM_DIRBROWSER | KM_FILEBROWSER | KM_FILTERSELECT | KM_TAGSELECT | KM_URLVIEW | KM_PODBOAT
+		KM_FEEDLIST | KM_ARTICLELIST | KM_ARTICLE | KM_DIALOGS | KM_DIRBROWSER | KM_FILEBROWSER | KM_FILTERSELECT | KM_TAGSELECT | KM_URLVIEW | KM_SEARCHRESULTSLIST |KM_PODBOAT
```
Comment (Minoru) [inline src/keymap.cpp:92]:
```suggestion
		KM_FEEDLIST | KM_ARTICLELIST | KM_ARTICLE | KM_DIALOGS | KM_DIRBROWSER | KM_FILEBROWSER | KM_FILTERSELECT | KM_TAGSELECT | KM_URLVIEW | KM_SEARCHRESULTSLIST | KM_PODBOAT
```

File: src/searchresultsformaction.cpp [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	RegexManager& r)
+	: ItemListFormAction(vv, formstr, cc, f, cfg, r) {};
+
+const std::vector<KeyMapHintEntry>& SearchResultsFormAction::get_keymap_hint() const
+{
+	static const std::vector<KeyMapHintEntry> hints = {
+		{OP_QUIT, _("Quit")},
+		{OP_OPEN, _("Open")},
+		{OP_PREVSEARCHRESULTS, _("Prev search results")}
+	};
```
Comment (Minoru) [inline src/searchresultsformaction.cpp:18]: I think this should also mention:
- the search key, so people understand that they can refine the search;
- the help key, just like the other dialogs do.

Also, I feel somewhat conflicted about contracting words. How about "Previous Results"?

File: src/keymap.cpp [hunk 6]
```diff
@@ -174,21 +174,21 @@ static const std::vector<OpDesc> opdescs = {
 		"toggle-article-read",
 		"N",
 		translatable("Toggle read status for article"),
-		KM_ARTICLELIST | KM_ARTICLE
+		KM_ARTICLELIST  | KM_SEARCHRESULTSLIST | KM_ARTICLE
```
Comment (Minoru) [inline src/keymap.cpp:177]:
```suggestion
		KM_ARTICLELIST | KM_SEARCHRESULTSLIST | KM_ARTICLE
```

File: src/keymap.cpp [hunk 7]
```diff
@@ -174,21 +174,21 @@ static const std::vector<OpDesc> opdescs = {
...
-		KM_FEEDLIST | KM_ARTICLELIST
+		KM_FEEDLIST | KM_ARTICLELIST | KM_SEARCHRESULTSLIST
 	},
 	{
 		OP_SHOWURLS,
 		"show-urls",
 		"u",
 		translatable("Show URLs in current article"),
-		KM_ARTICLE | KM_ARTICLELIST
+		KM_ARTICLE  | KM_SEARCHRESULTSLIST | KM_ARTICLELIST
```
Comment (Minoru) [inline src/keymap.cpp:191]:
```suggestion
		KM_ARTICLE | KM_SEARCHRESULTSLIST | KM_ARTICLELIST
```

File: src/keymap.cpp [hunk 8]
```diff
@@ -268,86 +268,86 @@ static const std::vector<OpDesc> opdescs = {
...
 	{
 		OP_EDITFLAGS,
 		"edit-flags",
 		"^E",
 		translatable("Edit flags"),
-		KM_ARTICLELIST | KM_ARTICLE
+		KM_ARTICLELIST | KM_SEARCHRESULTSLIST | KM_ARTICLE
 	},
-	{OP_NEXTFEED, "next-feed", "j", translatable("Go to next feed"), KM_ARTICLELIST},
+	{OP_NEXTFEED, "next-feed", "j", translatable("Go to next feed"), KM_ARTICLELIST | KM_SEARCHRESULTSLIST},
```
Comment (Minoru) [inline src/keymap.cpp:301]: I don't think this operation makes any sense for the search dialog: the feed that the search dialog shows is ephemeral, it was created just now to display a particular subset of articles. Thus, there is no "next feed" for it.

Same goes for `next-feed`, `prev-unread-feed`, and `next-unread-feed`.

File: src/keymap.cpp [hunk 9]
```diff
@@ -484,6 +484,14 @@ static const std::vector<OpDesc> opdescs = {
 	{OP_INT_SET, "set", "internal-set", "", KM_INTERNAL},
 
 	{OP_INT_GOTO_URL, "gotourl", "internal-goto-url", "", KM_INTERNAL},
+
+	{
```
Comment (Minoru) [inline src/keymap.cpp:488]: Please move this up before `OP_SK_*` and `OP_INT_*` ops.

File: src/keymap.cpp [hunk 10]
```diff
@@ -484,6 +484,14 @@ static const std::vector<OpDesc> opdescs = {
...
+		OP_PREVSEARCHRESULTS,
+		"prevsearchresults",
+		"z",
+		translatable("Previous search results"),
```
Comment (Minoru) [inline src/keymap.cpp:492]: This is an action, so let's describe it as one: "Return to previous search results (if any)".

File: src/keymap.cpp [hunk 11]
```diff
@@ -498,6 +506,7 @@ static const std::map<std::string, std::uint32_t> contexts = {
 	{"podboat", KM_PODBOAT},
 	{"dialogs", KM_DIALOGS},
 	{"dirbrowser", KM_DIRBROWSER},
+	{"searchresultslist", KM_SEARCHRESULTSLIST},
```
Comment (Minoru) [inline src/keymap.cpp:509]: This new context should be documented. I believe you can just git-grep for existing contexts, and that would lead you to all the places that might need updates.

File: src/searchresultsformaction.cpp [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+{
+	switch (op) {
+	case OP_PREVSEARCHRESULTS:
+		if (searchresultshistory.size() > 1) {
+			searchresultshistory.pop();
+			this->set_feed(searchresultshistory.top());
+		}
+		break;
+	default:
+		ItemListFormAction::process_operation(op, automatic, args);
```
Comment (Minoru) [inline src/searchresultsformaction.cpp:40]: I think this should say:

```suggestion
		return ItemListFormAction::process_operation(op, automatic, args);
```

Because if `ItemListFormAction` fails to process the operation, we want the caller to know about it.

File: src/searchresultsformaction.cpp [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+	Operation op,
+	bool automatic,
+	std::vector<std::string>* args)
+{
+	switch (op) {
+	case OP_PREVSEARCHRESULTS:
+		if (searchresultshistory.size() > 1) {
+			searchresultshistory.pop();
+			this->set_feed(searchresultshistory.top());
+		}
```
Comment (Minoru) [inline src/searchresultsformaction.cpp:37]: If the stack only has one element, there should be a message to the user explaining why the key does nothing.

File: include/searchresultsformaction.h [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+#include "rssfeed.h"
+#include "view.h"
+#include "keymap.h"
+#include <stack>
+
+namespace newsboat {
+
+class SearchResultsFormAction : public ItemListFormAction {
+public:
+	SearchResultsFormAction(View* vv,
```
Comment (Minoru) [inline include/searchresultsformaction.h:16]: It'd be nice if the constructor also took the initial feed to be added to the history, and also the searchphrase. That way, there's no need to call `set_feed()`, `add_history()`, and `set_searchphrase()` right after constructing the formaction, and then there's no way to screw things up by forgetting to call the methods.

File: src/view.cpp [hunk 15]
```diff
@@ -444,18 +445,26 @@ void View::push_searchresult(std::shared_ptr<RssFeed> feed,
...
+			searchresult->add_history(feed);
+			searchresult->init();
+			formaction_stack.push_back(searchresult);
+			current_formaction = formaction_stack_size() - 1;
+		} else {
+			auto searchresult = std::static_pointer_cast<SearchResultsFormAction>
+				(this->get_current_formaction());
+			searchresult->set_feed(feed);
+			searchresult->add_history(feed);
+			searchresult->set_searchphrase(phrase);
```
Comment (Minoru) [inline src/view.cpp:466]: I think these three methods should be combined into a single `add_to_history()` method, because they are always called simultaneously.

**Review by Minoru — CHANGES_REQUESTED** [review]
This is all the low-level feedback I got for now. I'll be using the feature for the next few days, and see if I spot anything else missing. Hopefully not, and then we can merge it!

Thanks once more for all the effort you're putting into this. Newsboat is not the simplest of projects, I'm glad you're not scared by its intricacies :)

File: include/itemlistformaction.h [hunk 16]
```diff
@@ -74,7 +74,7 @@ class ItemListFormAction : public ListFormAction {
 
 	void restore_selected_position();
 
-private:
+protected:
```
[the remark this answers is not in the document]
Comment (bogdasar1985) [reply include/itemlistformaction.h:77]: @Minoru Yes, I think we can mark only `process_operation` as `protected`.

File: include/view.h [hunk 17]
```diff
@@ -26,6 +26,7 @@ namespace newsboat {
 class ItemListFormAction;
 class ItemViewFormAction;
 class FeedListFormAction;
+class SearchResultsFormAction;
```
[the remark this answers is not in the document]
Comment (bogdasar1985) [reply include/view.h]: https://github.com/newsboat/newsboat/pull/2031/commits/03a65e8119e442be72e898f7316e9d9b98ce3442

File: include/itemlistformaction.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (bogdasar1985) [reply include/itemlistformaction.h:77]: https://github.com/newsboat/newsboat/pull/2031/commits/4513a238fb5df0051c25e084186fb097480b02fc

File: include/searchresultsformaction.h (hunk 2 above)
Comment (bogdasar1985) [reply include/searchresultsformaction.h]: https://github.com/newsboat/newsboat/pull/2031/commits/30cc562a54975d740a046c25d556d1e930d46e49

File: include/searchresultsformaction.h [hunk 18]
```diff
@@ -0,0 +1,43 @@
+#ifndef NEWSBOAT_SEARCHRESULTSFORMACTION_H_
+#define NEWSBOAT_SEARCHRESULTSFORMACTION_H_
+#include "configcontainer.h"
+#include "itemlistformaction.h"
+#include "logger.h"
+#include "regexmanager.h"
+#include "rssfeed.h"
+#include "view.h"
+#include "keymap.h"
+#include <stack>
```
[the remark this answers is not in the document]
Comment (bogdasar1985) [reply include/searchresultsformaction.h]: https://github.com/newsboat/newsboat/pull/2031/commits/210288609ad65d672b4fa5e22fa4ad5f9cb9b0a3

File: include/searchresultsformaction.h (hunk 1 above)
Comment (bogdasar1985) [reply include/searchresultsformaction.h]: https://github.com/newsboat/newsboat/pull/2031/commits/de6f4dce56fa5e8e9b21f22013d8fee4342af815

File: src/keymap.cpp (hunk 9 above)
Comment (bogdasar1985) [reply src/keymap.cpp:488]: @Minoru maybe even before `OP_CMD*` and `OP_OPEN*`?

File: src/itemlistformaction.cpp (hunk 3 above)
Comment (bogdasar1985) [reply src/itemlistformaction.cpp:919]: @Minoru Make an Issue for this. We'll discuss everything there.

File: include/searchresultsformaction.h (hunk 14 above)
Comment (bogdasar1985) [reply include/searchresultsformaction.h]: @Minoru but `ItemListFormAction` and other `FormAction` based classes don't do that. That consequence to inconsistent code and make a constructor too big.

File: src/view.cpp (hunk 15 above)
Comment (bogdasar1985) [reply src/view.cpp:466]: @Minoru I disagree, I think `set_feed` and `set_searchphrase`, can't be a part of area of responsibility of `add_to_history`

File: src/keymap.cpp (hunk 11 above)
Comment (bogdasar1985) [reply src/keymap.cpp:509]: @Minoru Do you mean add description in `doc/`? And if yes, why I need `git grep` for it?

File: src/itemlistformaction.cpp (hunk 3 above)
Comment (Minoru) [reply src/itemlistformaction.cpp:919]: https://github.com/newsboat/newsboat/issues/2036

File: src/keymap.cpp (hunk 9 above)
Comment (Minoru) [reply src/keymap.cpp:488]: Agreed, that'd look better.

File: src/keymap.cpp (hunk 11 above)
Comment (Minoru) [reply src/keymap.cpp:509]: Yeah, I mean `doc/`. `git grep` is the simplest way to find all the places where existing contexts are mentioned, and those are the places that probably need updating now.

File: include/searchresultsformaction.h (hunk 14 above)
Comment (Minoru) [reply include/searchresultsformaction.h]: Okay, let's keep it then. I can always change it later if I feel strongly about it :)

File: src/view.cpp (hunk 15 above)
Comment (Minoru) [reply src/view.cpp:466]: Why?

Note I called my method `add_to_history()`. The existing `add_history()` is a setter, so yeah, I agree it has a very small area of responsibility. But I think it's warranted to create an entirely new method, `add_to_history()`, which will combine the three and avoid temporal coupling between them. (`add_history()` can then be removed, because it won't be used anywhere anymore.)

**Review by Minoru — APPROVED** [review]
Looks good code-wise :+1: I'll try this out for a few more days, and either merge this, or provide more feedback.

File: doc/keycmds.dsv [hunk 19]
```diff
@@ -70,3 +70,4 @@ end||END||Go to the last item in the list.
 macro-prefix||,||Initiate macro execution. The next key press selects the actual macro and runs it.
 switch-focus||TAB||Switch focus between widgets. This is currently only applicable to the `filebrowser` and `dirbrowser` contexts.
 goto-title||||Go to item whose title contains the specified string (case-insensitive).
+prevsearchresults||z||Return to previous search results (if any). This only works from searchresultslist.
```
Comment (Minoru) [inline doc/keycmds.dsv:73]:
```suggestion
prevsearchresults||z||Return to previous search results (if any). This only works from `searchresultslist`.
```

(For consistency with other parts of this table.)

File: src/view.cpp (hunk 15 above)
Comment (Minoru) [reply src/view.cpp:466]: Ping.

Comment (bogdasar1985) [reply src/view.cpp]: @Minoru Done.

**Review by Minoru — APPROVED** [review]
Okay, I now used this for a bit and I don't see anything else that needs improving here, so let's finally merge!

Thank you once more for all the work, @bogdasar1985!
